# Dataset

Data Link : https://docs.google.com/spreadsheets/d/e/2PACX-1vQDzHCxv-dXHRjYoU24V2YFas4GrmHvbBiQeNHGLx7qoEGZCv_-ZAfXKZ_iURG3u11C0molRMPdq4Qw/pub?gid=1676240112&single=true&output=csv

```
data_link = "https://docs.google.com/spreadsheets/d/e/2PACX-1vQDzHCxv-dXHRjYoU24V2YFas4GrmHvbBiQeNHGLx7qoEGZCv_-ZAfXKZ_iURG3u11C0molRMPdq4Qw/pub?gid=1676240112&single=true&output=csv"
data = pd.read_csv(data_link)
```

In [ ]:
# Importing libraries
import pandas as pd
import numpy as np

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

#### First Look Dataset

In [ ]:
data_link = "https://docs.google.com/spreadsheets/d/e/2PACX-1vQDzHCxv-dXHRjYoU24V2YFas4GrmHvbBiQeNHGLx7qoEGZCv_-ZAfXKZ_iURG3u11C0molRMPdq4Qw/pub?gid=1676240112&single=true&output=csv"
data = pd.read_csv(data_link)
data.sample()

property_name property_type          society  price  \
3379  2 BHK Flat in Sector 68 Gurgaon          flat  pareena mi casa    1.1   

      price_per_sqft    area  \
3379          9020.0  1220.0   

                                           areaWithType  bedRoom  bathroom  \
3379  Super Built up area 1245(115.66 sq.m.)Carpet a...        2         2   

     balcony additionalRoom  \
3379       3  not available   

                                                address  floorNum facing  \
3379  1302, Tower - 3, Sector 68 Gurgaon, Gurgaon, H...      13.0   East   

        agePossession                                    nearbyLocations  \
3379  0 to 1 Year Old  ['Sector 55-56 Metro Station', 'Airia Mall', '...   

                                            description furnishDetails  \
3379  Residential apartment for sale in prominent lo...             []   

                                               features  \
3379  ['Security / Fire Alarm', 'Feng Shui / Vaastu ...   

                                                 rating  
3379  ['Environment4 out of 5', 'Lifestyle4 out of 5...

In [ ]:
# Checking No of instances and features
data.shape

(3961, 20)

* No of instances : 3961
* No of Features : 20

In [ ]:
# Info
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3961 entries, 0 to 3960
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   property_name    3961 non-null   object 
 1   property_type    3961 non-null   object 
 2   society          3960 non-null   object 
 3   price            3941 non-null   float64
 4   price_per_sqft   3941 non-null   float64
 5   area             3941 non-null   float64
 6   areaWithType     3961 non-null   object 
 7   bedRoom          3961 non-null   int64  
 8   bathroom         3961 non-null   int64  
 9   balcony          3961 non-null   object 
 10  additionalRoom   3961 non-null   object 
 11  address          3950 non-null   object 
 12  floorNum         3940 non-null   float64
 13  facing           2784 non-null   object 
 14  agePossession    3960 non-null   object 
 15  nearbyLocations  3754 non-null   object 
 16  description      3961 non-null   object 
 17  furnishDetails

* We have mixed data. Some columns are integer, float and object

In [ ]:
# Describe numeric columns
data.describe().T

count          mean            std    min     25%     50%  \
price           3941.0      2.475499       2.922570   0.07     0.9     1.5   
price_per_sqft  3941.0  13915.092870   23181.160459   2.00  6790.0  9000.0   
area            3941.0   4647.025628  117585.647103  45.00  1187.0  1700.0   
bedRoom         3961.0      3.363040       2.009067   1.00     2.0     3.0   
bathroom        3961.0      3.406968       2.049265   1.00     2.0     3.0   
floorNum        3940.0      6.667766       6.005669   0.00     2.0     4.0   

                     75%        max  
price               2.65       31.5  
price_per_sqft  13750.00   600000.0  
area             2290.00  7250000.0  
bedRoom             4.00       36.0  
bathroom            4.00       36.0  
floorNum           10.00       51.0

In [ ]:
# Getting percentage of null values in each columns
data.isna().sum(axis=0)/data.shape[0] * 100

property_name       0.000000
property_type       0.000000
society             0.025246
price               0.504923
price_per_sqft      0.504923
area                0.504923
areaWithType        0.000000
bedRoom             0.000000
bathroom            0.000000
balcony             0.000000
additionalRoom      0.000000
address             0.277708
floorNum            0.530169
facing             29.714719
agePossession       0.025246
nearbyLocations     5.225953
description         0.000000
furnishDetails     26.054027
features           17.899520
rating             11.360767
dtype: float64

* There are null values in our data we need to pre-processed it

In [ ]:
# Checking for duplicates data samples
data.duplicated().sum()

0

* NO duplicates samples

In [ ]:
# Looking at data samples which have missing values
data[data.isna().any(axis=1)].shape

(2171, 20)

* Out of 3961 rows, there are 2171 rows which have some missing data.

In [ ]:
#
data.isna().sum(axis=1).value_counts()

0    1790
1    1181
2     600
3     288
4      87
5      14
6       1
dtype: int64

In [ ]:
data[data.isna().sum(axis=1)==5].head(3)

property_name property_type  \
317           5 Bedroom House for sale in Laxman Vihar         house   
421           2 Bedroom House for sale in Shakti Nagar         house   
678  1 Bedroom House for sale in Hno 13 Sagar encla...         house   

            society  price  price_per_sqft   area  \
317  not applicable    NaN             NaN    NaN   
421     independent    NaN             NaN    NaN   
678     independent   0.85          8910.0  954.0   

                    areaWithType  bedRoom  bathroom balcony additionalRoom  \
317  Plot area 670(560.21 sq.m.)        5         3       2  not available   
421   Plot area 1000(92.9 sq.m.)        2         2       1  not available   
678   Plot area 106(88.63 sq.m.)        1         2       0  not available   

                                         address  floorNum facing  \
317               Laxman Vihar, Gurgaon, Haryana       3.0    NaN   
421               Shakti Nagar, Gurgaon, Haryana       1.0    NaN   
678  Hno 13 Sagar enclave Sura, Gurgaon, Haryana       1.0    NaN   

        agePossession                                    nearbyLocations  \
317  5 to 10 Year Old  ['Chintapurni Mandir', 'State bank ATM', 'Prat...   
421  5 to 10 Year Old  ['Lal Superspeciality Hospital', 'Shri Gobind ...   
678   1 to 5 Year Old                                                NaN   

                                           description  \
317  I have a house/villa available for sale in sar...   
421  2 bath, dimensions 1000 x 500 ft, semi-Furnish...   
678      Independent house for sale finishing required   

                                        furnishDetails features  \
317                                                 []      NaN   
421  ['1 Light', 'No AC', 'No Bed', 'No Chimney', '...      NaN   
678                                                NaN      NaN   

                                                rating  
317  ['Environment4 out of 5', 'Lifestyle4 out of 5...  
421  ['Environment4 out of 5', 'Lifestyle4 out of 5...  
678                                                NaN

In [ ]:
# Checking Property Name
data.property_name[:5]

0      3 BHK Flat in Sector 36 Sohna
1    2 BHK Flat in Sector 89 Gurgaon
2    2 BHK Flat in Sector 86 Gurgaon
3                2 BHK Flat in Sohna
4    3 BHK Flat in Sector 92 Gurgaon
Name: property_name, dtype: object

* Property name has info about Sector and BHK

In [ ]:
# Checking Property Type
data.property_type.unique()

array(['flat', 'house'], dtype=object)

* Type of the property : either flat or house

In [ ]:
# Checking Society
# data.society.unique()
data.society.nunique()

725

* There are many societies, too much for a category.

In [ ]:
# Checking Price Column
data.price.describe()

count    3941.000000
mean        2.475499
std         2.922570
min         0.070000
25%         0.900000
50%         1.500000
75%         2.650000
max        31.500000
Name: price, dtype: float64

* Price is in Crore

In [ ]:
# Checking Price Column
data.areaWithType[0],data.areaWithType.sample(1).values[0]

('Super Built up area 1081(100.43 sq.m.)Carpet area: 650 sq.ft. (60.39 sq.m.)',
 'Plot area 360(301.01 sq.m.)')

* This column have the different area type with value

In [ ]:
# Number of different types rooms : edroom, bathroom, balcony, additional room
data[['bedRoom', 'bathroom', 'balcony','additionalRoom']].sample(2)

bedRoom  bathroom balcony additionalRoom
350        2         2       2  not available
781        4         4       2  not available

In [ ]:
data.additionalRoom = data.additionalRoom.str.lower()

In [ ]:
set(data.additionalRoom[:10].str.cat(sep=',').split(','))

{'not available',
 'others',
 'pooja room',
 'servant room',
 'store room',
 'study room'}

* Balcony columns has data values like 3+. Need to pre-process
* additionalRoom have mulivalued data which can be transfered into columns

In [ ]:
# Checking floorNum column
data.floorNum.unique()

array([ 2.,  4.,  0., 17., 10.,  5., 20., 14.,  8.,  1.,  3., 16.,  9.,
        7., 31., 12., 45., 15.,  6., 19., 11., 18., 41., 29., 23., 13.,
       27., 32., 44., 21., 22., 30., nan, 25., 24., 28., 39., 35., 26.,
       34., 43., 33., 51., 38., 40.])

In [ ]:
data.floorNum.value_counts()

3.0     552
2.0     547
1.0     404
4.0     339
8.0     200
7.0     188
6.0     188
10.0    187
5.0     181
9.0     171
12.0    162
0.0     139
14.0    120
11.0    117
15.0     83
16.0     54
17.0     51
13.0     44
18.0     38
19.0     34
20.0     26
21.0     16
23.0     15
22.0     14
24.0     11
25.0      9
27.0      8
30.0      5
26.0      5
31.0      4
33.0      4
28.0      4
29.0      3
34.0      2
35.0      2
43.0      2
40.0      2
39.0      2
32.0      2
44.0      1
45.0      1
51.0      1
38.0      1
41.0      1
Name: floorNum, dtype: int64

* floorNum meaning
  - in flat type property : floor number of flat, like on 3rd floor or so
  - in house type property : total number of floors in the house

In [ ]:
# Checking facing column
data.facing.unique()

array([nan, 'North-West', 'North-East', 'East', 'North', 'South',
       'South-West', 'West', 'South-East'], dtype=object)

In [ ]:
data.facing.value_counts()

East          668
North-East    663
North         410
West          258
South         237
North-West    206
South-East    183
South-West    159
Name: facing, dtype: int64

In [ ]:
data.facing.isna().sum()

1177

* Facing columns has many missing values, but have some prominent description about the property

In [ ]:
# Checking agePossession
data.agePossession.unique()

array(['0 to 1 Year Old', 'Within 6 months', 'Under Construction',
       '1 to 5 Year Old', '10+ Year Old', '5 to 10 Year Old', 'undefined',
       'Aug 2024', 'May 2024', 'Nov 2023', 'Jul 2027', 'Oct 2024',
       'Dec 2023', 'By 2025', 'Jun 2025', 'Mar 2026', 'Apr 2024',
       'Jan 2025', 'Dec 2024', 'By 2023', 'By 2024', 'Dec 2025',
       'Jul 2026', 'Oct 2025', 'Jun 2024', 'Nov 2025', 'Mar 2024',
       'By 2027', 'Within 3 months', 'Feb 2024', 'Aug 2025', 'Apr 2026',
       'Jan 2024', 'Aug 2023', 'Oct 2023', 'Jul 2024', 'Sep 2023',
       'Sep 2025', 'Mar 2025', 'Jun 2027', 'Jan 2027', 'Jan 2026',
       'Nov 2024', 'Jan 2028', nan, 'Jul 2025', 'Dec 2026', 'Aug 2026',
       'Sep 2027'], dtype=object)

* Need preprocessing : Have data about since when this propety is being possessed or will be available for possession.

In [ ]:
# Checking nearbyLocations
data.nearbyLocations[0]

"['Sector 55-56 Metro Station', 'The Leaf Mall Sohna', 'Omaxe Gurgaon Mall', 'Badshahpur Sohna Rd Hwy', 'Vidya Niketan School', 'KIIT College of Engg', 'Scottish High International School', 'Rawal Institutions', 'Park Hospital', 'Indira Gandhi Intl Airport', 'Spaze Business Park', 'DLF Cyber City']"

In [ ]:
nearbyLocations_counts = pd.Series(data.nearbyLocations.str.strip('[|]').str.cat(sep=', ').split(', ')).apply(lambda x: x.strip("'\"")).value_counts()
nearbyLocations_counts[:100]

Hdfc bank                              1129
Indira Gandhi International Airport    1083
Dwarka Expressway                       703
Icici bank ATM                          581
Pizza Hut                               549
                                       ... 
Kidney Clinic                           155
Hdfc ATM                                154
Nirula's                                154
Kamla International                     154
SPAZE BUSINESS PARK                     154
Length: 100, dtype: int64

In [ ]:
'dr.','eye centre', 'dental', 'Ortho','Optho'

('dr.', 'eye centre', 'dental', 'Ortho', 'Optho')

* Have multivalues pointsbut very informtive in respect to a property.
* This could affect the prices.
* Need processing
* We will create different columns from here like: Malls, Stations, Airport, Educational Institute, Hospitals, Banks,

In [ ]:
# Checking furnishDetails columns
data.furnishDetails.isna().sum()

1032

In [ ]:
# Group by 'group' and count missing values in 'value'
# Defining a custom function to count missing values and non null data counts
def count_missing_and_total(x):
    return pd.Series({'missing_count': x.isna().sum(), 'total_count': x.count()})

# Group by 'property_type' and apply the custom function
data.groupby('property_type')['furnishDetails'].agg(count_missing_and_total).reset_index()

property_type furnishDetails
0          flat    [797, 2200]
1         house     [235, 729]

* Too much missing data


In [ ]:
# Checking on features -(menaing amenties) column
amenties_value_counts = pd.Series(data.features.str.strip('[|]').str.cat(sep=', ').split(', ')).apply(lambda x: x.strip("'\"")).value_counts()
amenties_value_counts

Park                               2887
Maintenance Staff                  2572
Lift(s)                            2456
Visitor Parking                    2427
Water Storage                      2313
Feng Shui / Vaastu Compliant       2087
Intercom Facility                  2058
Security / Fire Alarm              1964
Fitness Centre / GYM               1928
Club house / Community Center      1910
Swimming Pool                      1825
Rain Water Harvesting              1776
Security Personnel                 1501
Waste Disposal                     1448
Shopping Centre                    1436
Natural Light                      1294
No open drainage around            1248
Internet/wi-fi connectivity        1240
Airy Rooms                         1199
High Ceiling Height                1179
Piped-gas                          1106
Separate entry for servant room    1056
Water softening plant              1046
Low Density Society                 938
Spacious Interiors                  910


In [ ]:
# Checking ratings columns
data['rating']

0                                                     NaN
1       ['Environment4 out of 5', 'Lifestyle4 out of 5...
2       ['Environment5 out of 5', 'Safety4.5 out of 5'...
3       ['Safety4 out of 5', 'Lifestyle4 out of 5', 'E...
4       ['Green Area5 out of 5', 'Construction5 out of...
                              ...                        
3956    ['Environment5 out of 5', 'Lifestyle4 out of 5...
3957                                                  NaN
3958    ['Environment4 out of 5', 'Lifestyle4.5 out of...
3959                                                  NaN
3960    ['Environment4 out of 5', 'Lifestyle4 out of 5...
Name: rating, Length: 3961, dtype: object

In [ ]:
# Group by 'group' and count missing values in 'value'
# Defining a custom function to count missing values and non null data counts
def count_missing_and_total(x):
    return pd.Series({'missing_count': x.isna().sum(), 'total_count': x.count()})

# Group by 'property_type' and apply the custom function
data.groupby('property_type')['rating'].agg(count_missing_and_total).reset_index()

property_type       rating
0          flat  [329, 2668]
1         house   [121, 843]

In [ ]:
data['rating'][1]

"['Environment4 out of 5', 'Lifestyle4 out of 5', 'Connectivity4 out of 5', 'Safety4 out of 5']"

* Rating given on 4 parameters : Envinonment, Lifestyle, Connectivity, Safety

### Conclusion based on pre-limenary alaysis (First data look)

******
Features to be created from
******
* Nearby Locations
  - Airport
  - num_malls
  - num_hospitals
  - railway_station
  - num_banks
  - num_educational_institutes

* Amenties/features
  - hasLift
  - hasSwimingPool
  - Security / Fire Alarm
  - Fitness Centre / GYM

## Preprocessing

#### Creating Sector Comlumn

In [ ]:
data.property_name[:5].values

array(['3 BHK Flat in Sector 36 Sohna', '2 BHK Flat in Sector 89 Gurgaon',
       '2 BHK Flat in Sector 86 Gurgaon', '2 BHK Flat in Sohna',
       '3 BHK Flat in Sector 92 Gurgaon'], dtype=object)

In [ ]:
# Extracting sector numbers

data['sector'] = data['property_name'].str.extract(r'Sector (\d+)')

data['sector'].isna().sum()

1014

In [ ]:
# Extracting sector number from address column
data['sector_ad'] = data['address'].str.extract(r'Sector (\d+)')

In [ ]:
data['sector_ad'].isna().sum()

952

In [ ]:
data[data[['sector', 'sector_ad']].isna().all(axis=1)].shape

(946, 22)

* There are some missing values in sector column which can be imputed from address columns (sector_ad) data and vice versa (for 6 rows).

In [ ]:
# Filling sector columns data from sector_ad columns
data['sector'].fillna(data['sector_ad'], inplace=True)
data['sector'].isna().sum()

946

In [ ]:
data.drop('sector_ad', axis=1, inplace=True)

* May be we can find data from internet or scrapping for remaing sector based on the address column or LATITUDE or LONGITUDE

#### Creating features from nearby loactions

Nearby Locations

* Airport
* num_malls
* num_hospitals
* railway_station
* num_banks
* num_educational_institutes



In [ ]:
data['airport'] = data['nearbyLocations'].str.lower().str.contains('airport')
data['station'] = data['nearbyLocations'].str.lower().str.contains('station|junction')

In [ ]:
data['nearbyLocations'].isna().sum()

207

In [ ]:
data['airport']

0        True
1       False
2        True
3       False
4        True
        ...  
3956     True
3957    False
3958    False
3959      NaN
3960    False
Name: airport, Length: 3961, dtype: object

In [ ]:
# PyCaret - to remove outliers

In [ ]:
# data['hospitals'] = data['nearbyLocations'].str.lower().str.contains('dr. |eye centre| dental| Ortho| Optho|clinic| hospital')

data['hospitals'] = data['nearbyLocations'].str.lower().str.count('dr\. |eye centre| dental| ortho| optho|clinic| hospital|medical| nursing|aiims')


In [ ]:
# data['hospitals']
data[['nearbyLocations','hospitals']].tail(10)

nearbyLocations  hospitals
3951  ['Dlf phase 1 metro station', 'Sector 42-43 me...        7.0
3952  ['Global City Centre', 'Sohna Road', 'Damdama ...        1.0
3953  ['MG Road Metro Station', 'NeoSquare Shopping ...        1.0
3954  ['Sheetla Mata Mandir', 'Chintapurni Mandir', ...       45.0
3955  ['Dwarka Expressway', 'RPS International Schoo...        3.0
3956  ['Dwarka sector 21 metro station', 'Pacific D2...        1.0
3957  ['Palam Vihar Vyapar kendra', 'Palam triangle'...        3.0
3958  ['Iffco chowk metro station', 'Huda city centr...        6.0
3959                                                NaN        NaN
3960  ['Airia Mall', 'Sohna Road', 'Vidyamrat Public...        1.0

In [ ]:

data['banks'] = data['nearbyLocations'].str.lower().str.count('bank')

In [ ]:
data[['banks','nearbyLocations']].head(10)

banks                                    nearbyLocations
0    0.0  ['Sector 55-56 Metro Station', 'The Leaf Mall ...
1    0.0  ['Sant Soordas Sihi Metro Station', 'Airia Mal...
2    0.0  ['V3S Sapphire Ninety Mall', 'Sector 86 Road',...
3    1.0  ['Huda City Centre', 'Omaxe City Centre', 'Nin...
4    0.0  ['Sapphire 93 Mall', 'Dwarka Expressway', 'NH-...
5    0.0  ['The Hive', 'JMS Marine Square', 'Gurugram Rd...
6    0.0  ['Sector 55-56 Metro Station', 'Ambience Mall'...
7    0.0  ['Galleria 108 Mall', 'Dwarka Expressway', 'Ce...
8    0.0  ['WTC Plaza', 'Star Mall', 'Dwarka Expy, Dhanw...
9    4.0  ['Palam Vihar Vyapar kendra', 'Chintapurni Man...

In [ ]:
data['malls']= data['nearbyLocations'].str.lower().str.count('mall|city centre|city mart|plaza|shopping complex | cyberhub')

In [ ]:
data[['malls','nearbyLocations']].tail(10)

malls                                    nearbyLocations
3951    0.0  ['Dlf phase 1 metro station', 'Sector 42-43 me...
3952    1.0  ['Global City Centre', 'Sohna Road', 'Damdama ...
3953    1.0  ['MG Road Metro Station', 'NeoSquare Shopping ...
3954    0.0  ['Sheetla Mata Mandir', 'Chintapurni Mandir', ...
3955    0.0  ['Dwarka Expressway', 'RPS International Schoo...
3956    1.0  ['Dwarka sector 21 metro station', 'Pacific D2...
3957    2.0  ['Palam Vihar Vyapar kendra', 'Palam triangle'...
3958    1.0  ['Iffco chowk metro station', 'Huda city centr...
3959    NaN                                                NaN
3960    1.0  ['Airia Mall', 'Sohna Road', 'Vidyamrat Public...

In [ ]:
data['educational_institutes']= data['nearbyLocations'].str.lower().str.count('university|college|institutes|academy|coaching| school| library')

In [ ]:
data[['educational_institutes','nearbyLocations']].tail(10)

educational_institutes  \
3951                     2.0   
3952                     2.0   
3953                     1.0   
3954                     3.0   
3955                     5.0   
3956                     2.0   
3957                     3.0   
3958                     2.0   
3959                     NaN   
3960                     1.0   

                                        nearbyLocations  
3951  ['Dlf phase 1 metro station', 'Sector 42-43 me...  
3952  ['Global City Centre', 'Sohna Road', 'Damdama ...  
3953  ['MG Road Metro Station', 'NeoSquare Shopping ...  
3954  ['Sheetla Mata Mandir', 'Chintapurni Mandir', ...  
3955  ['Dwarka Expressway', 'RPS International Schoo...  
3956  ['Dwarka sector 21 metro station', 'Pacific D2...  
3957  ['Palam Vihar Vyapar kendra', 'Palam triangle'...  
3958  ['Iffco chowk metro station', 'Huda city centr...  
3959                                                NaN  
3960  ['Airia Mall', 'Sohna Road', 'Vidyamrat Public...

#### Crating features from Amenties
* Amenties/features
  - hasLift
  - hasSwimingPool
  - Security / Fire Alarm
  - Fitness Centre / GYM
  - has park
  - parking
  - water storage

In [ ]:
data['has_lift']=data['features'].str.lower().str.contains('lift')

In [ ]:
data['has_swimming_pool']=data['features'].str.lower().str.contains('pool')

In [ ]:
data['has_security_alarm']=data['features'].str.lower().str.contains('alarm')

In [ ]:
data['has_gym']=data['features'].str.lower().str.contains('gym |fitness ')

In [ ]:
data['has_park']=data['features'].str.lower().str.contains('park')

In [ ]:
data['has_parking']=data['features'].str.lower().str.contains(' parking')

In [ ]:
data['has_water_storage']=data['features'].str.lower().str.contains('water storage')

In [ ]:
data[['features','has_lift','has_swimming_pool','has_security_alarm','has_gym','has_park','has_parking','has_water_storage']].tail(10)

features has_lift  \
3951  ['Feng Shui / Vaastu Compliant', 'Private Gard...    False   
3952  ['Power Back-up', 'Feng Shui / Vaastu Complian...    False   
3953  ['Centrally Air Conditioned', 'Water purifier'...    False   
3954                                                NaN      NaN   
3955  ['Security / Fire Alarm', 'Power Back-up', 'Fe...    False   
3956  ['Water purifier', 'Centrally Air Conditioned'...    False   
3957  ['Power Back-up', 'Intercom Facility', 'Lift(s...    False   
3958  ['Private Garden / Terrace', 'Water purifier',...    False   
3959                                                NaN      NaN   
3960  ['Security / Fire Alarm', 'Power Back-up', 'Fe...    False   

     has_swimming_pool has_security_alarm has_gym has_park has_parking  \
3951              True              False    True     True        True   
3952              True              False    True     True       False   
3953              True               True    True     True        True   
3954               NaN                NaN     NaN      NaN         NaN   
3955              True               True   False     True        True   
3956              True               True    True     True        True   
3957              True              False    True     True       False   
3958             False              False    True    False       False   
3959               NaN                NaN     NaN      NaN         NaN   
3960              True               True    True     True        True   

     has_water_storage  
3951              True  
3952             False  
3953              True  
3954               NaN  
3955              True  
3956              True  
3957             False  
3958              True  
3959               NaN  
3960              True

In [ ]:
data.head(1)

property_name property_type                  society  \
0  3 BHK Flat in Sector 36 Sohna          flat  signature global park 4   

   price  price_per_sqft    area  \
0   0.82          7585.0  1081.0   

                                        areaWithType  bedRoom  bathroom  \
0  Super Built up area 1081(100.43 sq.m.)Carpet a...        3         2   

  balcony  ... banks malls  educational_institutes has_lift has_swimming_pool  \
0       2  ...   0.0   2.0                     3.0    False             False   

  has_security_alarm has_gym has_park has_parking has_water_storage  
0              False   False     True       False             False  

[1 rows x 34 columns]

In [ ]:
data['furnishDetails'].head(10)

0                                                  NaN
1                                                  NaN
2                                                   []
3                                                   []
4    ['3 Wardrobe', '6 Fan', '1 Exhaust Fan', '3 Ge...
5    ['3 Fan', '10 Light', 'No AC', 'No Bed', 'No C...
6    ['4 Wardrobe', '3 Fan', '1 Geyser', '1 Microwa...
7    ['5 AC', 'No Bed', 'No Chimney', 'No Curtains'...
8    ['7 Fan', '4 Geyser', '6 Light', '5 AC', 'No B...
9                                                  NaN
Name: furnishDetails, dtype: object

In [ ]:
# data.to_csv('gurgain_properties_19-09.csv', index=False)

### TO-DO (19-09)
* PRICE COLUMN CLEANING [Done]
- AREA CLEANING [Done]
- AGE POSSESSION [Done]
- RATING COLUMN [Done]
- Furnish Details Column [Done]
- Dropping Extra columns [Done]


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np
data_path = "/content/drive/MyDrive/PG Thappar - CSE/Data Science Foundation/DSF-Labs/Real-Estate/Data/gurgaon_properties_19-09-sector_na_602.csv"

In [ ]:
data = pd.read_csv(data_path)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3961 entries, 0 to 3960
Data columns (total 35 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   property_name           3961 non-null   object 
 1   property_type           3961 non-null   object 
 2   society                 3960 non-null   object 
 3   price                   3941 non-null   float64
 4   price_per_sqft          3941 non-null   float64
 5   area                    3941 non-null   float64
 6   areaWithType            3961 non-null   object 
 7   bedRoom                 3961 non-null   int64  
 8   bathroom                3961 non-null   int64  
 9   balcony                 3961 non-null   object 
 10  additionalRoom          3961 non-null   object 
 11  address                 3950 non-null   object 
 12  floorNum                3940 non-null   float64
 13  facing                  2784 non-null   object 
 14  agePossession           3960 non-null   

In [ ]:
data.head(1)

property_name property_type                  society  \
0  3 BHK Flat in Sector 36 Sohna          flat  signature global park 4   

   price  price_per_sqft    area  \
0   0.82          7585.0  1081.0   

                                        areaWithType  bedRoom  bathroom  \
0  Super Built up area 1081(100.43 sq.m.)Carpet a...        3         2   

  balcony  ... malls educational_institutes  has_lift has_swimming_pool  \
0       2  ...   2.0                    3.0     False             False   

  has_security_alarm has_gym has_park has_parking has_water_storage SECTOR  
0              False   False     True       False             False     36  

[1 rows x 35 columns]

#### Price Columns Cleaning

In [ ]:
data[data['price'].isna()]

property_name property_type  \
39      5 Bedroom House for sale in Sector 66 Gurgaon         house   
317          5 Bedroom House for sale in Laxman Vihar         house   
421          2 Bedroom House for sale in Shakti Nagar         house   
621     4 Bedroom House for sale in Sector 66 Gurgaon         house   
849     9 Bedroom House for sale in Sector 45 Gurgaon         house   
961     5 Bedroom House for sale in Sector 66 Gurgaon         house   
1474    5 Bedroom House for sale in Sector 33 Gurgaon         house   
1539          5 Bedroom House for sale in DLF Phase 2         house   
2056                  2 BHK Flat in Sector 93 Gurgaon          flat   
2096          5 Bedroom House for sale in DLF Phase 3         house   
2226    5 Bedroom House for sale in Sector 33 Gurgaon         house   
2291   5 Bedroom House for sale in Sector 108 Gurgaon         house   
2560        7 Bedroom House for sale in Naharpur Rupa         house   
3065    6 Bedroom House for sale in Sector 11 Gurgaon         house   
3135    5 Bedroom House for sale in Sector 33 Gurgaon         house   
3165         16 Bedroom House for sale in DLF Phase 4         house   
3502  4 Bedroom House for sale in Udyog Vihar Phase 1         house   
3605    5 Bedroom House for sale in Sector 66 Gurgaon         house   
3622         12 Bedroom House for sale in DLF Phase 2         house   
3625      5 Bedroom House for sale in Nirvana Country         house   

                                society  price  price_per_sqft  area  \
39                   emaar mgf marbella    NaN             NaN   NaN   
317                      not applicable    NaN             NaN   NaN   
421                         independent    NaN             NaN   NaN   
621                  emaar mgf marbella    NaN             NaN   NaN   
849              unitech greenwood city    NaN             NaN   NaN   
961                  emaar mgf marbella    NaN             NaN   NaN   
1474           unitech uniworld resorts    NaN             NaN   NaN   
1539                        independent    NaN             NaN   NaN   
2056  signature global orchard avenue 2    NaN             NaN   NaN   
2096              dlf city plot phase 3    NaN             NaN   NaN   
2226           unitech uniworld resorts    NaN             NaN   NaN   
2291                         sobha city    NaN             NaN   NaN   
2560                        independent    NaN             NaN   NaN   
3065                        independent    NaN             NaN   NaN   
3135           unitech uniworld resorts    NaN             NaN   NaN   
3165                        independent    NaN             NaN   NaN   
3502                        independent    NaN             NaN   NaN   
3605                 emaar mgf marbella    NaN             NaN   NaN   
3622                        independent    NaN             NaN   NaN   
3625                     unitech espace    NaN             NaN   NaN   

                                           areaWithType  bedRoom  bathroom  \
39                          Plot area 350(292.64 sq.m.)        5         5   
317                         Plot area 670(560.21 sq.m.)        5         3   
421                          Plot area 1000(92.9 sq.m.)        2         2   
621   Plot area 350(292.64 sq.m.)Built Up area: 6500...        4         4   
849                         Plot area 240(200.67 sq.m.)        9         9   
961                         Plot area 350(292.64 sq.m.)        5         5   
1474                        Plot area 502(419.74 sq.m.)        5         7   
1539  Plot area 502(419.74 sq.m.)Built Up area: 450 ...        5         5   
2056                  Carpet area: 577.11 (53.62 sq.m.)        2         2   
2096                 Built Up area: 5000 (464.52 sq.m.)        5         5   
2226                        Plot area 520(434.79 sq.m.)        5         8   
2291                 Built Up area: 7331 (681.07 sq.m.)        5         1   
2560                 

In [ ]:
data.dropna(subset=['price'], inplace=True)

* Nan values removed : We dropped all rows with `nan` values in price column

In [ ]:
data.price.unique()

array([ 0.82,  0.95,  0.46,  0.32,  1.6 ,  0.48,  7.5 ,  2.  ,  1.9 ,
        1.2 ,  4.5 ,  0.4 ,  1.55,  2.25,  2.77,  1.09,  1.15,  5.5 ,
        1.77,  0.85,  0.9 ,  0.6 ,  9.  ,  7.  ,  5.65,  2.15,  4.4 ,
        0.64,  2.86,  3.5 ,  0.7 ,  1.54,  0.45,  8.25,  6.1 ,  2.1 ,
        2.4 ,  1.25,  2.8 ,  2.58,  3.  ,  2.45,  1.38,  1.71,  1.85,
        0.81,  5.6 ,  1.5 ,  1.45,  1.65,  0.35,  5.75,  1.12,  0.75,
        0.25,  0.5 ,  0.67,  0.56,  1.75,  4.15,  3.4 ,  2.3 ,  7.6 ,
        2.99,  3.87,  2.65,  9.3 ,  1.19,  1.4 ,  1.  ,  1.3 ,  1.36,
        1.1 , 16.  , 10.45,  0.34,  0.39,  3.65,  1.29,  0.54,  0.79,
       10.  ,  0.28,  7.49,  1.44,  2.37,  0.2 ,  1.05,  1.83,  3.2 ,
        6.95,  4.55,  1.72,  2.2 ,  0.63,  1.13,  1.95,  4.7 ,  6.5 ,
        0.55,  1.78,  2.5 ,  3.9 ,  1.67,  0.65,  3.12,  2.05, 11.  ,
        1.41,  0.98,  4.35,  1.59,  2.28,  1.21,  0.93,  2.24,  0.78,
        1.93,  2.6 ,  7.46,  4.8 ,  0.97,  1.7 ,  6.75,  0.69,  1.69,
        8.5 ,  5.  ,

In [ ]:
data.price.dtype

dtype('float64')

#### AREA CLEANING

In [ ]:
# Number of nan values
data.area.isna().sum()

0

In [ ]:
# Data tyes of Area column
data.area.dtype

dtype('float64')

In [ ]:
# Unique Values
data.area.unique()

array([ 1081.,  1105., 58228., ...,  3042.,  5490.,   788.])

In [ ]:
# Unique Values
data.area.describe()

count    3.941000e+03
mean     4.647026e+03
std      1.175856e+05
min      4.500000e+01
25%      1.187000e+03
50%      1.700000e+03
75%      2.290000e+03
max      7.250000e+06
Name: area, dtype: float64

#### AGE POSSESSION

In [ ]:
# Checking Nan Vlaues
data.agePossession.isna().sum()

1

In [ ]:
data[data.agePossession.isna()]

property_name property_type         society  price  \
2461  2 BHK Flat in Sector 79 Gurgaon          flat  m3m golfestate    1.4   

      price_per_sqft    area                      areaWithType  bedRoom  \
2461         10000.0  1400.0  Carpet area: 1400 (130.06 sq.m.)        2   

      bathroom balcony  ... malls educational_institutes  has_lift  \
2461         2       2  ...   0.0                    0.0     False   

     has_swimming_pool has_security_alarm has_gym has_park has_parking  \
2461              True              False    True     True       False   

     has_water_storage SECTOR  
2461             False     79  

[1 rows x 35 columns]

In [ ]:
# Seeing unique values
data.agePossession.unique()

array(['0 to 1 Year Old', 'Within 6 months', 'Under Construction',
       '1 to 5 Year Old', '10+ Year Old', '5 to 10 Year Old', 'undefined',
       'Aug 2024', 'May 2024', 'Nov 2023', 'Jul 2027', 'Oct 2024',
       'Dec 2023', 'By 2025', 'Jun 2025', 'Mar 2026', 'Apr 2024',
       'Jan 2025', 'Dec 2024', 'By 2023', 'By 2024', 'Dec 2025',
       'Jul 2026', 'Oct 2025', 'Jun 2024', 'Nov 2025', 'Mar 2024',
       'By 2027', 'Within 3 months', 'Feb 2024', 'Aug 2025', 'Apr 2026',
       'Jan 2024', 'Aug 2023', 'Oct 2023', 'Jul 2024', 'Sep 2023',
       'Sep 2025', 'Mar 2025', 'Jun 2027', 'Jan 2027', 'Jan 2026',
       'Nov 2024', 'Jan 2028', nan, 'Jul 2025', 'Dec 2026', 'Aug 2026',
       'Sep 2027'], dtype=object)

In [ ]:


# Define a function to categorize values
def categorize_age(age):
    if pd.isna(age):
        return 'undefined'
    elif 'Old' in age  in age or pd.to_datetime(age, errors='coerce') < pd.to_datetime('Aug 2023', format='%b %Y', errors='coerce'):
        return 'constructed'
    else:
        return 'underConstruction'

data['status'] = data['agePossession'].apply(categorize_age)
data[['status', 'agePossession']]

status       agePossession
0           constructed     0 to 1 Year Old
1     underConstruction     Within 6 months
2     underConstruction  Under Construction
3           constructed     0 to 1 Year Old
4           constructed     1 to 5 Year Old
...                 ...                 ...
3956        constructed     1 to 5 Year Old
3957        constructed    5 to 10 Year Old
3958        constructed    5 to 10 Year Old
3959        constructed     1 to 5 Year Old
3960        constructed     1 to 5 Year Old

[3941 rows x 2 columns]

* Created a column `status` which have two values : `constructed and underConstruction`

#### RATING COLUMN

In [ ]:
# Checking Nan values in rating column
data.rating.isna().sum()

448

In [ ]:
data.rating[0:5]


0                                                  NaN
1    ['Environment4 out of 5', 'Lifestyle4 out of 5...
2    ['Environment5 out of 5', 'Safety4.5 out of 5'...
3    ['Safety4 out of 5', 'Lifestyle4 out of 5', 'E...
4    ['Green Area5 out of 5', 'Construction5 out of...
Name: rating, dtype: object

In [ ]:
x = data.rating.str.strip('\[|]').str.cat(sep=', ').split(', ')
pd.Series(x).str.strip('\'').str.extract(r'([A-Za-z\s]+)\d+\sout').value_counts()

Connectivity    2938
Environment     1699
Lifestyle       1671
Safety          1519
Amenities       1451
Green Area      1415
Management      1376
Construction    1335
dtype: int64

In [ ]:
def calculate_average_rating(row):
  ratings = []
  # print(row, type(row))
  if type(row)!=float:
    for value in row.strip('\[|\]').split(', '):
      # print(value)
      rating = int(value.split(' out')[0][-1])
      ratings.append(rating)
    if ratings:
      return sum(ratings) / len(ratings)
    else:
      return None


# Apply the function to create a new column with average ratings
data['avgRating'] = data['rating'].apply(calculate_average_rating)

* Created a rating columns based on averaging all different parameter ratings

In [ ]:
data.head(2)

property_name property_type                  society  \
0    3 BHK Flat in Sector 36 Sohna          flat  signature global park 4   
1  2 BHK Flat in Sector 89 Gurgaon          flat         smart world gems   

   price  price_per_sqft    area  \
0   0.82          7585.0  1081.0   
1   0.95          8600.0  1105.0   

                                        areaWithType  bedRoom  bathroom  \
0  Super Built up area 1081(100.43 sq.m.)Carpet a...        3         2   
1                   Carpet area: 1103 (102.47 sq.m.)        2         2   

  balcony  ... has_lift has_swimming_pool  has_security_alarm has_gym  \
0       2  ...    False             False               False   False   
1       2  ...    False              True               False    True   

  has_park has_parking has_water_storage SECTOR             status avgRating  
0     True       False             False     36        constructed       NaN  
1     True       False             False     89  underConstruction       4.0  

[2 rows x 37 columns]

In [ ]:
data.facing.value_counts()

East          665
North-East    661
North         408
West          255
South         236
North-West    205
South-East    181
South-West    159
Name: facing, dtype: int64

In [ ]:
data.furnishDetails.isna().sum()

1026

#### FurnishDetails

In [ ]:
import ast

# Define the keys_list
keys_list = ['Modular Kitchen', 'Bed', 'Washing Machine', 'TV', 'Stove', 'Sofa', 'Fridge', 'Microwave', 'Dining Table', 'Water Purifier', 'Wardrobe', 'Curtains', 'Chimney', 'AC', 'Light', 'Geyser', 'Exhaust Fan', 'Fan']

# Function to create a dictionary from the input string
def create_furnishing_dict(input_string):
    item_dict = {key: 0 for key in keys_list}
    if type(input_string)!=float:
      input_list = ast.literal_eval(input_string)
      for item in input_list:
          v = item.split(' ', 1)[0]
          key = item.split(' ', 1)[1]
          item_dict[key]=v
      return item_dict
    else:
      return None
data['furnishDetailsDict'] = data.furnishDetails.apply(create_furnishing_dict)

In [ ]:
def get_furnish(x):
  if x and type(x)!=float:
    if col in x:
      return x[col]
    else:
      return 0
  else:
    return None

keys_list = ['Modular Kitchen', 'Bed', 'Washing Machine', 'TV', 'Stove', 'Sofa', 'Fridge', 'Microwave', 'Dining Table', 'Water Purifier', 'Wardrobe', 'Curtains', 'Chimney', 'AC', 'Light', 'Geyser', 'Exhaust Fan', 'Fan']
for col in keys_list:
  data[col] = data.furnishDetailsDict.apply(get_furnish)

In [ ]:
data.shape

(3941, 56)

In [ ]:
data[keys_list+['furnishDetailsDict']].head(10)

Modular Kitchen   Bed Washing Machine    TV Stove  Sofa Fridge Microwave  \
0            None  None            None  None  None  None   None      None   
1            None  None            None  None  None  None   None      None   
2               0     0               0     0     0     0      0         0   
3               0     0               0     0     0     0      0         0   
4               1    No              No    No    No    No     No        No   
5              No    No              No    No    No    No     No        No   
6               1    No              No    No     1    No     No         1   
7              No    No              No    No    No    No     No        No   
8              No    No              No    No    No    No     No        No   
9            None  None            None  None  None  None   None      None   

  Dining Table Water Purifier Wardrobe Curtains Chimney    AC Light Geyser  \
0         None           None     None     None    None  None  None   None   
1         None           None     None     None    None  None  None   None   
2            0              0        0        0       0     0     0      0   
3            0              0        0        0       0     0     0      0   
4           No             No        3        4       1     5    15      3   
5           No             No       No       No      No    No    10     No   
6           No             No        4        5       1     7    10      1   
7           No             No       No       No      No     5    No     No   
8           No             No       No       No      No     5     6      4   
9         None           None     None     None    None  None  None   None   

  Exhaust Fan   Fan                                 furnishDetailsDict  
0        None  None                                               None  
1        None  None                                               None  
2           0     0  {'Modular Kitchen': 0, 'Bed': 0, 'Washing Mach...  
3           0     0  {'Modular Kitchen': 0, 'Bed': 0, 'Washing Mach...  
4           1     6  {'Modular Kitchen': '1', 'Bed': 'No', 'Washing...  
5          No     3  {'Modular Kitchen': 'No', 'Bed': 'No', 'Washin...  
6          No     3  {'Modular Kitchen': '1', 'Bed': 'No', 'Washing...  
7          No    No  {'Modular Kitchen': 'No', 'Bed': 'No', 'Washin...  
8          No     7  {'Modular Kitchen': 'No', 'Bed': 'No', 'Washin...  
9        None  None                                               None

#### Dropping Columns

In [ ]:
# Looking all columns
data.columns

Index(['property_name', 'property_type', 'society', 'price', 'price_per_sqft',
       'area', 'areaWithType', 'bedRoom', 'bathroom', 'balcony',
       'additionalRoom', 'address', 'floorNum', 'facing', 'agePossession',
       'nearbyLocations', 'description', 'furnishDetails', 'features',
       'rating', 'sector', 'airport', 'station', 'hospitals', 'banks', 'malls',
       'educational_institutes', 'has_lift', 'has_swimming_pool',
       'has_security_alarm', 'has_gym', 'has_park', 'has_parking',
       'has_water_storage', 'SECTOR', 'status', 'avgRating',
       'furnishDetailsDict', 'Modular Kitchen', 'Bed', 'Washing Machine', 'TV',
       'Stove', 'Sofa', 'Fridge', 'Microwave', 'Dining Table',
       'Water Purifier', 'Wardrobe', 'Curtains', 'Chimney', 'AC', 'Light',
       'Geyser', 'Exhaust Fan', 'Fan'],
      dtype='object')

In [ ]:
data.to_csv('/content/drive/MyDrive/PG Thappar - CSE/Data Science Foundation/DSF-Labs/Real-Estate/Data/More_Data/Gurgaon_data_with_furnishing_ratings_20-09-bckp.csv', index=False)

In [ ]:
data.drop(columns=['agePossession', 'price_per_sqft', 'sector', 'areaWithType','furnishDetails', 'features','furnishDetailsDict', 'rating', 'description', 'nearbyLocations'], axis=1, inplace=True)

In [ ]:
data.to_csv('/content/drive/MyDrive/PG Thappar - CSE/Data Science Foundation/DSF-Labs/Real-Estate/Data/Gurgaon_property_data_final_v0_for_EDA.csv', index=False)
# Saving Done

In [ ]:
data.head()

### To Do (20-09) (on hold):
- Deal with Society
- Deal with Address
- Deal with Additional Rooms
- Facing (Filling Nan Values)

In [ ]:
data.head(2)

property_name property_type                  society  \
0    3 BHK Flat in Sector 36 Sohna          flat  signature global park 4   
1  2 BHK Flat in Sector 89 Gurgaon          flat         smart world gems   

   price    area  bedRoom  bathroom balcony           additionalRoom  \
0   0.82  1081.0        3         2       2            not available   
1   0.95  1105.0        2         2       2  study room,servant room   

                                       address  floorNum facing airport  \
0       144, Sector 36 Sohna, Gurgaon, Haryana       2.0    NaN    True   
1  O 118d, Sector 89 Gurgaon, Gurgaon, Haryana       4.0    NaN   False   

  station  hospitals  banks  malls  educational_institutes has_lift  \
0    True        1.0    0.0    2.0                     3.0    False   
1    True        1.0    0.0    1.0                     2.0    False   

  has_swimming_pool has_security_alarm has_gym has_park has_parking  \
0             False              False   False     True       False   
1              True              False    True     True       False   

  has_water_storage SECTOR             status  avgRating Modular Kitchen  \
0             False     36        constructed        NaN            None   
1             False     89  underConstruction        4.0            None   

    Bed Washing Machine    TV Stove  Sofa Fridge Microwave Dining Table  \
0  None            None  None  None  None   None      None         None   
1  None            None  None  None  None   None      None         None   

  Water Purifier Wardrobe Curtains Chimney    AC Light Geyser Exhaust Fan  \
0           None     None     None    None  None  None   None        None   
1           None     None     None    None  None  None   None        None   

    Fan  
0  None  
1  None